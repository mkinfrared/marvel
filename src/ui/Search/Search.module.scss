.Search {
  position: relative;
  width: 300px;

  .dropdown {
    position: absolute;
    top: calc(100% + 5px);
    width: 100%;
  }
}
