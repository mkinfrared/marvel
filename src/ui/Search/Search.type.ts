export interface SearchProps<T> {
  className?: string;
  getOptionLabel: (option: T) => string;
  loading?: boolean;
  onChange?: (option: T) => void;
  onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  options?: T[];
  value: string;
}
