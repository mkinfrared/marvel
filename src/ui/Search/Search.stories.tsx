import { Meta, Story } from "@storybook/react/types-6-0";
import React, { useState } from "react";

import { SearchProps } from "ui/Search/Search.type";

import { Search } from "./Search";

export default {
  title: "UI/Search",
  component: Search
} as Meta;

type Option = {
  id: number;
  name: string;
};

const options = [
  { id: 1, name: "Thor" },
  { id: 2, name: "Natasha" },
  { id: 3, name: "Deadpool" }
];

const Template: Story<SearchProps<Option>> = (args) => {
  const [inputValue, setInputValue] = useState("");
  const [selectedOption, setSelectedOption] = useState<Option | null>(null);

  return (
    <>
      <Search
        {...args}
        value={inputValue}
        onInputChange={(event) => setInputValue(event.target.value)}
        onChange={(option) => setSelectedOption(option)}
        getOptionLabel={(option) => option.name}
      />
      <p style={{ marginTop: 20 }}>Selected option: {selectedOption?.name}</p>
    </>
  );
};

const Primary = Template.bind({});

Primary.args = {
  options,
  getOptionLabel: (option) => option.name
};

export { Primary };
