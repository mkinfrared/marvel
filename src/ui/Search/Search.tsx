import React, { useCallback, useMemo, useState } from "react";

import Dropdown from "ui/Dropdown";
import Input from "ui/Input";
import classes from "utils/classes";

import css from "./Search.module.scss";
import { SearchProps } from "./Search.type";

const Search = <T,>({
  className,
  getOptionLabel,
  onChange,
  onInputChange,
  value,
  options = [],
  loading = false
}: SearchProps<T>) => {
  const [dropdownVisible, setDropdownVisible] = useState(false);

  const handleFocus = useCallback(() => {
    setDropdownVisible(true);
  }, []);

  const handleBlur = useCallback(() => {
    setDropdownVisible(false);
  }, []);

  const showDropdown = dropdownVisible && (options.length > 0 || loading);

  const handleChange = useCallback(
    (option: T) => () => {
      const label = getOptionLabel(option);
      const event = { target: { value: label } };

      onChange && onChange(option);

      onInputChange(event as React.ChangeEvent<HTMLInputElement>);
    },
    []
  );

  // const getOptions = () => {
  //   return options.map((option, index) => {
  //     const { id } = option as Record<string, any>;
  //     const key = id ?? index;
  //
  //     if (renderOptions) {
  //       return (
  //         <React.Fragment key={key}>{renderOptions(option)}</React.Fragment>
  //       );
  //     }
  //
  //     return (
  //       <li key={key} onMouseDown={handleChange(option)}>
  //         {getOptionLabel(option)}
  //       </li>
  //     );
  //   });
  // };

  const optionsToRender = useMemo(() => {
    return options.map((option, index) => {
      const { id } = option as Record<string, any>;
      const key = id ?? index;

      return (
        <li key={key} onMouseDown={handleChange(option)}>
          <p>{getOptionLabel(option)}</p>
        </li>
      );
    });
  }, [getOptionLabel, handleChange, options]);

  return (
    <div className={classes(css.Search, className)}>
      <Input
        onChange={onInputChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        value={value}
      />
      {showDropdown && (
        <Dropdown loading={loading} className={css.dropdown}>
          {optionsToRender}
        </Dropdown>
      )}
    </div>
  );
};

export { Search };

export default React.memo(Search) as typeof Search;
