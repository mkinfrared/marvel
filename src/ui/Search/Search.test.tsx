import { fireEvent, render } from "@testing-library/react";
import React from "react";

import { Search } from "./Search";

describe("<Search />", () => {
  const value = "marklar";
  const options = [{ name: "foobar" }];
  const getOptionLabel = jest.fn(({ name }) => name);
  const onInputChange = jest.fn();
  const onChange = jest.fn();
  const props = { value, getOptionLabel, onInputChange, options, onChange };
  const Component = <Search {...props} />;

  it("should be defined", () => {
    const { container } = render(Component);

    expect(container).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should render <Dropdown /> when input is focused and options have length greater than 0", () => {
    const { container } = render(Component);
    const input = container.querySelector("input");

    let dropdown = container.querySelector(".dropdown");

    expect(dropdown).toBeNull();

    fireEvent.focus(input!);

    dropdown = container.querySelector(".dropdown");

    expect(dropdown).toBeDefined();
  });

  it("should render <Dropdown /> when input is focused and loading equals to true", () => {
    const Comp = <Search {...props} loading options={undefined} />;
    const { container } = render(Comp);
    const input = container.querySelector("input");

    let dropdown = container.querySelector(".dropdown");

    expect(dropdown).toBeNull();

    fireEvent.focus(input!);

    dropdown = container.querySelector(".dropdown");

    expect(dropdown).toBeDefined();
  });

  it("should hide <Dropdown /> when input is blurred", () => {
    const Comp = <Search {...props} loading options={undefined} />;
    const { container } = render(Comp);
    const input = container.querySelector("input");

    fireEvent.focus(input!);

    let dropdown = container.querySelector(".dropdown");

    expect(dropdown).toBeDefined();

    fireEvent.blur(input!);

    dropdown = container.querySelector(".dropdown");

    expect(dropdown).toBeNull();
  });

  it("should set input's value to clicked option text", () => {
    const { container } = render(Component);
    const input = container.querySelector("input");

    fireEvent.focus(input!);

    const option = container.querySelector("li");

    expect(option).toBeDefined();

    fireEvent.mouseDown(option!);

    expect(onInputChange).toHaveBeenCalledTimes(1);

    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
