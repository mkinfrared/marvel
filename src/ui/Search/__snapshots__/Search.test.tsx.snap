// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should match the snapshot 1`] = `
<div>
  <div
    class="Search"
  >
    <div
      class="Input"
    >
      <input
        type="text"
        value="marklar"
      />
    </div>
  </div>
</div>
`;
